src/services/pcn-packetcapture/src/Packetcapture.cpp

```diff
@@ -165,23 +173,22 @@ void Packetcapture::writeDump(const std::vector<uint8_t> &packet){
   struct timeval tp;
   std::chrono::system_clock::duration tp_dur = (timeP + temp_offset).time_since_epoch();
   to_timeval(tp_dur, tp);
-  
+
   p->setTimestampSeconds((uint32_t) tp.tv_sec);
   p->setTimestampMicroseconds((uint32_t) tp.tv_usec);
   p->setPacketlen((uint32_t) packet.size());
-  if (p->getPacketlen() > filters->getSnaplen()) {
-    p->setCapturelen(filters->getSnaplen());
+  if (p->getPacketlen() > getSnaplen()) {
+    p->setCapturelen(getSnaplen());
   } else {
     p->setCapturelen(p->getPacketlen());
   }
   p->setRawPacketData(packet);
 
   struct pcap_pkthdr *pkt_hdr = new struct pcap_pkthdr;
-  pkt_hdr->ts_sec = p->getTimestampSeconds();
-  pkt_hdr->ts_usec = p->getTimestampMicroseconds();
+  pkt_hdr->ts = tp;
```

Comment:
Hi @DavideAG, thanks for the review

My problem here was that in the library pcap.h (that i need for pcap function for the filter) is defined the same struct but with other variable, in particolar here is the def. in pcap.h:
```
struct pcap_pkthdr {
	struct timeval ts;	/* time stamp */
	bpf_u_int32 caplen;	/* length of portion present */
	bpf_u_int32 len;	/* length this packet (off wire) */
};
```
here yours:
```
struct pcap_pkthdr {
	bpf_u_int32 ts_sec;	
        bpf_u_int32 ts_usec;	
	bpf_u_int32 caplen;	/* length of portion present */
	bpf_u_int32 len;	/* length this packet (off wire) */
};
```
but struct timeval is defined as:
```
struct timeval {
  __time_t tv_sec;		/* Seconds.  */
  __suseconds_t tv_usec;	/* Microseconds.  */
};
```
so I think is the same,  I only joined the two things
everything else is the same as your code

Reply:
I see your problem.
I tried your solution too, but I had problems when I was dumping packages directly into the server's file system (so without using the packetcapture client written in python).

Were you able to read correctly with wireshark a dump written directly from packetcapture? It should be stored in the tmp folder.
In my case, using the Linux struct timeval, the dumps produced by the packetcapture in the server file system were not readable.

I finally found a solution regarding the correct way to dump packages into the pcap file.
The problem is related to the byte size of the struct timeval variables.
These must be forced to 32bit as suggested by wireshark wiki.
https://wiki.wireshark.org/Development/LibpcapFileFormat#Record_.28Packet.29_Header


However, regardless of the structure you use, **the packet timestamp must be acquired in the fast path** of the service and **not in the slow path** otherwise it will not be accurate.

Reply:
No I'm not able, when I try to open the dump file in tmp folder wireshark say:

> The capture file appears to be damaged or corrupt.
> (pcap: File has 639023-byte packet, bigger than maximum of 262144)

Reply:
I've changed the code and now I take the packet timestamp in the fast path as you also done